### ISSUE TITLE
Proof of work[NIP-13] Positive Interoperability Tracker

### ISSUE THREAD

**Issue by alltheseas** [nostrability/nostrability#103]
There are those who advocate POW may deter spam on nostr. Now that replyguy era is here, it is time to verify. 

**app**|**Nip-13**|**Metric A**|**B**|**C**|**comment**
-----|-----|-----|-----|-----|-----|-----|-----
damus|❌|?|?| | | | 
Wherostr|✅|?|?| | | | 
coracle|?|?|?| | | | 
nostrudel|?|?|.| | | |

**Comment by alexgleason** [comment]
ReplyGuy is doing proof of work. Example: https://njump.me/nevent1qqsqqqrfkwpcjsev4zw9vw8t9d6q3srazmmak68d6t6qvq9zwl759ccprdmhxue69uhkwmr9v9ek7mnpw3hhytnyv4mz7un9d3shjevzzxa

**Comment by alltheseas** [comment]
@alexgleason are there any apps that overcome replyguy's POW spam? 

If yes, for how long?

**Comment by alltheseas** [comment]
Do relays check for POW before accepting an event?

**Comment by alexgleason** [comment]
@alltheseas In Ditto I started requiring the user to have a valid NIP-05 name OR sign events with a proof-of-work difficulty of 24 or higher (ReplyGuy is using difficulty of 17) and this has stopped the spam for now.

Proof of work can limit the sheer volume of spam from dozens per second to 2-3 per second on the same hardware, but 2-3 per second and the spammer is still happy. If you crank the difficulty up so it takes 5 seconds per message, honestly the spammer is still decently happy with this while users suffer. Therefore proof of work is "nice to have", but I think the barrier is more of a knowledge and accessibility one that deters only the most unmotivated script kiddies.

What @vitorpamplona says here is true: https://njump.me/nevent1qqsy8msnju5hxcv48n7xxknvyjld3psz9ymjw0kqzuuqqj9q3hj7efqprdmhxue69uhkwmr9v9ek7mnpw3hhytnyv4mz7un9d3shjtryn0g

**Comment by alltheseas** [comment]
> pow is for discovery, not spam

https://njump.me/note1t5g7ht4cw96j76femagp4a4tl59jjxdhyk30sfrnfavw9sun78ssah722g

**Comment by alltheseas** [comment]
Primal example of top POW note feed

https://m.primal.net/KwvQ.png

https://njump.me/note1dcd8acfu9gt4zpa7d3htkjtgfzrf0pdaql5l402h4c7praq3u7qs367un5

**Comment by alltheseas** [comment]
Added getwired and ditto

**Comment by alexgleason** [comment]
Look at how much is blocked by my policy of requiring NIP-05 or PoW: https://njump.me/nevent1qqsq0kj7eag9c45qcsjg9x9h9mhdr5h9q3key48n35jyaa2rlgny3hqprdmhxue69uhkwmr9v9ek7mnpw3hhytnyv4mz7un9d3shjdklzh9

To be fair I think this is mostly actual spam, but there are for sure some casualties.

**Comment by alltheseas** [comment]
Wherostr has filter by pow and create pow implemented

![image](https://github.com/user-attachments/assets/1487458a-6258-46d0-a1fc-8e94606a72ee)

![image](https://github.com/user-attachments/assets/6cf2029a-5d76-416f-b28d-29a4ef9b8aea)

**Comment by vitorpamplona** [comment]
The only option is WoT. Other stuff can serve as a deterrent but in the end everything (NIP05, rate limits, IP limits, pow, hidden words, block lists) can be "easily" circumvented. 

And when I mean WoT, I mean true WoT, not the shallow follows of follows filter that everybody is calling WoT. 

Put pubkeys in a pagerank-style graph and you will see where the humans and the bots are.

**Comment by alexgleason** [comment]
@vitorpamplona I'm still not sure what to do as a relay. The "WoT" relays I've seen are all seeded with one guy's profile. The only thing that makes sense is to track a pubkey's behavior over time, and limit what they can do outside of what looks like "normal" behavior.

**Comment by pippellia-btc** [comment]
> The only option is WoT. Other stuff can serve as a deterrent but in the end everything (NIP05, rate limits, IP limits, pow, hidden words, block lists) can be "easily" circumvented.

I agree @vitorpamplona , with the only exception that PoW I believe can work if externalized to a service provider, because this means there is a profit-incentive to have ASICs which level the playing field even against very sophisticated attackers.

The issue with PoW is more the UX, as "pay as u post" is pretty bad from the pov of the user. I think there is a lot of value in [PoW endorsements / PoW keys](https://pippellia.com/pippellia/Social+Graph/Navigating+the+social+graph#PoW%20keys) in this regard, because it's a one time cost that has a tangible benefit (pay 10$ and get a PoW checkmark displayed on clients).

> Put pubkeys in a pagerank-style graph and you will see where the humans and the bots are.

Fuck yeah. However, this can't be done client side. I am working on a service that does these kind of computations which also maintains some level of client side verification.

**Comment by alexgleason** [comment]
Can yall explain exactly what you mean by "pagerank-style graph"? How does that work in practice?

**Comment by vitorpamplona** [comment]
> Can yall explain exactly what you mean by "pagerank-style graph"?

Each pubkey is a node and holds a score between 0 to 1. Every `a`, `e` and `p` tag is a graph edge that represents a "vote" between those nodes. But not all votes are equal. The value of the vote depends on the score of who is casting the vote. Initially, every pubkey is given the same score (often normalized to 1). Then, the algorithm recalculates the rank of each pubkey by summing all the votes to it. The algorithm normalizes all scores to 1. And repeats this process until all scores stabilize. 

Today, a simple score cut-off filters out most of the spam. New users will need to ask friends to like their content to be included in the graph. 

But, as spammers figure this out, they will build bot swarms with strong scores between themselves. 

When that time arrives, something like modularity maximization (Louvain Algorithm) or spectral clustering can be used to separate the charts into communities and then you can use the fact that the user is inside one of these communities to lower the score of the rest.

Anyway, this can only be done on relays. 

Clients can do a lighter version with their local databases, giving more weight to the things each user sees daily. A user that is unknown but keeps replying to you is better than new users coming into the user's graph.

**Comment by alexgleason** [comment]
I spent the last few hours reading about pagerank algorithm and arguing with chatgpt about it. Then ReplyGuy came back, and I just want to say that blocking nip05 domains does work. https://gleasonator.dev/@alex@gleasonator.dev/posts/340ae36fa69896f9ee2538b8027a4d342017103049e24c97c9f8945b36585e3f

**Comment by alltheseas** [comment]
Where is/are the spam nip 
05 list/s 😅

**Comment by vitorpamplona** [comment]
I couldn't resist. 

NIP-99
======

Spam
------

`draft` `optional`

This NIP defines an event with `kind:65535` that contains only spam. Bots can use it to boost any agenda they prefer. 

Contents can be literally anything, in any format, in any language. Clients SHOULD use NIP-10 rules to build threads and notify other pubkey keys. 

Example:

```json
{
  "kind": 65535
  "tags": [
    ["e", <event-id>, <relay-url>, "<marker>"],
    ["alt", "this is just spam"]
  ],
  "content": "the spam"
}
```

**Comment by alexgleason** [comment]
@vitorpamplona You forgot a `p` tag in that example.

**Comment by nostrband** [comment]
Pagerank-style thing works. On nostr.band we know of 136k replyguy profiles, but only 2k managed to get non-zero rank, and if we cared enough to ban top-ranked spammers manually then those 2k would be down to 200 or less. 

A couple notes:

> Initially, every pubkey is given the same score (often normalized to 1).

That would be a mistake, leading to:

> But, as spammers figure this out, they will build bot swarms with strong scores between themselves.

We call our thing "trust rank" because we don't give same score to everyone at the start. Instead, we give initial weight to select set of "trusted" pubkeys, which means that bot farms don't get any to start with. Then as the weights spread bot farms only get anything leaked to them if someone trusted reacts to bot's content, which is a rare occasion compared to the amount of spam, and can be alleviated by manual moderation. 

Without this initial trusted set we would end up with all weight concentrated among bot farms, not real users. Google started with some manually curated website directory for their PageRank, without it you'd get "what's popular" (bot farm!), not "what's popular and high quality". 

> How does that work in practice?

It works really well if you have good coverage like nostr.band that crawls all relays. But I'm not sure how that would work if the data set is limited to the local database of a relay or a client. My guess would be that the number of false-positives (non-spam marked as spam) would skyrocket and would be very unpredictable. More research should be done here to figure things out. I'm more hopeful about the client-side for spam filtering here - we'd only need current user's graph and it's surroundings which would probably have good coverage in the local client's db. A relay that has to serve everyone and needs global trust rank would probably perform poorly with only local data.

> I agree @vitorpamplona , with the only exception that PoW I believe can work if externalized to a service provider

So users would be paying to the service provider for pow. Why pay for pow if you can pay to relays directly and not waste the cpu to boil the oceans? PoW seems to make very little sense in the long-term here. I wonder if something like an ecash-zap-to-relay event would work better - it's as easy to verify as pow for relays and clients, but funds relays directly, instead of wasting computation on leading zeroes. 

Re. pagerank, I'm happy to provide a bounty and a server and data to someone to experiment and make a good scalable open-source solution suitable for relays and clients. Would you be interested @pippellia-btc ?

**Comment by staab** [comment]
I don't like the idea of using PageRank widely because it's a global view of the network. Aren't we just re-building the legacy web at this point? Nostr's strength is subjective trust, which is what I'd like to lean into more. This can be augmented with trusted nip05s, trusted relays, trusted badge issuers, etc.

**Comment by brugeman** [comment]
> I don't like the idea of using PageRank widely because it's a global view of the network. Aren't we just re-building the legacy web at this point? Nostr's strength is subjective trust, which is what I'd like to lean into more. This can be augmented with trusted nip05s, trusted relays, trusted badge issuers, etc.

It's just a tool like many other. It can have many uses and many variations, so open source research here feels valuable to make this tool more practically usable. Like with search relays, clients could just let you select PR providers if it turns out PR actually solves some problems well.  

Also, "legacy web" is decentralized and scaled pretty damn well, I wouldn't equate it to centralized social media.

**Comment by vitorpamplona** [comment]
> This can be augmented with trusted nip05s, trusted relays, trusted badge issuers, etc.

Sure, any extra information can be added to these types of algorithms. I used the word "pagerank" just because devs generally know what the algo is about and it does get the point across. But there are dozens, if not hundreds, of options out there to explore. In the end, it's all the same thing: Graph analysis. 

In the long term, all clients will have to code a version of WoT Graph analysis for their local cache. There is no way around it.

**Comment by pippellia-btc** [comment]
> So users would be paying to the service provider for pow. Why pay for pow if you can pay to relays directly and not waste the cpu to boil the oceans?

fair critic @brugeman @nostrband . That's why I am not suggesting to pay per each note, but pay once for a PoW badge/PoW key/PoW endorsement, which is a social signal, and a product because "want to protect your audience against impersonators? get your PoW checkmark!". Trustless, one-time social signal whose work can be externalized.

> PoW seems to make very little sense in the long-term here. I wonder if something like an ecash-zap-to-relay event would work better - it's as easy to verify as pow for relays and clients, but funds relays directly, instead of wasting computation on leading zeroes.

There is no way to make this work in a trustless way, because any proof "Alice paid Bob" is useless if Alice and Bob are colluding, for example if Bob gives back the money to Alice in a private and non traceable way.
So you would have to trust Alice and Bob with their reputation, but this is a circular argument because we are trying to determine their reputation.

> Re. pagerank, I'm happy to provide a bounty and a server and data to someone to experiment and make a good scalable open-source solution suitable for relays and clients. Would you be interested?

I am developing something similar, in the form of a paid service, because the thing is that it requires computational resources, it needs to be up-to-date (imagine if a popular npub gets compromised), and it's just a lot of work to maintain. Packaging it inside a relay would make relays quite expensive to run, so I am not sure it's a good idea, but I am happy to chat more about it.

> I don't like the idea of using PageRank widely because it's a global view of the network. Aren't we just re-building the legacy web at this point? Nostr's strength is subjective trust, which is what I'd like to lean into more. This can be augmented with trusted nip05s, trusted relays, trusted badge issuers, etc.

@staab the way I see it is that Pagerank/Trustrank can be used when there is no particular viewpoint, for example in the case of free relays (e.g. Damus) to do optimized rate limiting.
High-rank? you can write 100 times a minute
Low-rank? you can write just 5 times a minute, sorry.
Shouldn't just Damus offer a paid relay then? maybe, idk, but I think that's another argument.

However, when there is a viewpoint, for example a user looking for recommended npubs to follow, then I agree with you and the client should use personalized pagerank, which is a flavour of pagerank that gives initial weight to the user.

**Spoiler alert**: personalized pagerank (pp) and trustrank are actually the same thing. They are called in different ways because they have been used in different scenarios.

With trustrank, the initial weight is distributed to a large set of "reputable npubs".
With pp, the initial weight is all concentrated on the user.
They are the same because of an important property, meaning the linearity of pp.
The pp of me AND you, is the average of my pp and your pp.

More here: https://ieeexplore.ieee.org/stamp/stamp.jsp?tp=&arnumber=8895775

**Comment by alltheseas** [comment]
Lume UI MacOS

![Image](https://github.com/user-attachments/assets/da0c6747-75cb-47fe-8799-94d2733b56b0)
